README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# mifa - multiple imputation for factor analysis

<img src='man/figures/logo.png' align="right" height="139" />

<!-- badges: start -->

[![Lifecycle:
maturing](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)
[![R build
status](https://github.com/teebusch/mifa/workflows/R-CMD-check/badge.svg)](https://github.com/teebusch/mifa/actions)
[![Codecov test
coverage](https://codecov.io/gh/Teebusch/mifa/branch/master/graph/badge.svg)](https://codecov.io/gh/Teebusch/mifa?branch=master)

<!-- badges: end -->

`mifa` is an R package that implements multiple imputation of covariance
matrices to allow to perform factor analysis on incomplete data. It
works as follows:

1.  Impute missing values multiple times using *Multivariate Imputation
    with Chained Equations* (MICE) from the
    [mice](https://amices.org/mice/) package.

2.  Combine the covariance matrices of the imputed data sets into a
    single covariance matrix using Rubin’s rules<sup>[1]</sup>

3.  Use the combined covariance matrix for exploratory factor analysis.

`mifa` also provides two types of confidence intervals for the variance
explained by different numbers of principal components: Fieller
confidence intervals (parametric) for larger samples<sup>[2]</sup> and
bootstrapped confidence intervals (nonparametric) for smaller
samples.<sup>[3]</sup>

**For more information about the method, see:**

Nassiri, V., Lovik, A., Molenberghs, G., Verbeke, G. (2018). On using
multiple imputation for exploratory factor analysis of incomplete data.
*Behavior Research Methods* 50, 501–517. doi:
[10.3758/s13428-017-1013-4](https://doi.org/10.3758/s13428-017-1013-4)

*Note:* The paper was accompanied by an implementation in R, and this
package emerged from it. The repository appears to have been abandoned
by the authors, but you can still find it
[here](https://github.com/vahidnassiri/mifa).

## Installation

You can install `mifa` from [GitHub](https://github.com/teebusch/mifa)
with:

``` r
# install.packages("devtools")
devtools::install_github("teebusch/mifa")
```

## Usage

### Example Data

For this example we use the `bfi` data set from the `psych` package. It
contains 2,800 subjects’ answers to 25 personality self-report items and
3 demographic variables (sex, education, and age). Each of the 25
personality questions is meant to tap into one of the “Big 5”
personality factors, as indicated by their names: **O**penness,
**C**onscientiousness, **A**greeableness, , **E**xtraversion,
**N**euroticism. There are missing responses for most items. Instead of
dropping the incomplete cases from the analysis, we will use `mifa` to
impute them, and then perform a factor analysis on the imputed
covariance matrix.

### Imputing the Covariance Matrix

First, we use `mifa()` to impute the covariance matrix and get an idea
how many factors we should use. We use the `cov_vars` argument to tell
`mifa` to use `gender`, `education`, and `age` for the imputations, but
exclude them from the covariance matrix:

``` r
library(mifa)
library(psych)

mi <- mifa(
  data      = bfi, 
  cov_vars  = -c(gender, education, age),
  n_pc      = 2:8, 
  ci        = "fieller", 
  print     = FALSE
)

mi
#> Imputed covariance matrix of 25 variables
#> 
#> Variable:   A1 A2 A3 A4 A5 C1 C2 C3 C4 C5 E1 E2 E3 E4 E5 N1 N2 N3 N4 N5 O1 O2 O3 O4 O5 
#> N Imputed:  16 27 26 19 16 21 24 20 26 16 23 16 25  9 21 22 21 11 36 29 22  0 28 14 20
#> 
#> Number of MICE imputations: 5
#> Additional variables used for imputations:
#> gender education age
#> 
#> Cumulative proportion of variance explained by n principal components:
#> 
#> n  prop  Fieller CI    
#> 2  0.33  [0.32, 0.34]  
#> 3  0.41  [0.40, 0.42]  
#> 4  0.48  [0.47, 0.49]  
#> 5  0.54  [0.53, 0.55]  
#> 6  0.59  [0.58, 0.59]  
#> 7  0.62  [0.61, 0.63]  
#> 8  0.66  [0.65, 0.66]
```

### Factor Analysis

It looks like the first 5 principal components explain more than half of
the variance in the responses, so we perform a factor analysis with 5
factors, using the `fa()` function from the `psych` package. We can get
the imputed covariance matrix of our data from `mi$cov_combined`. From
there on, it’s business as usual.

``` r
fit <- fa(mi$cov_combined, n.obs = nrow(bfi), nfactors = 5)
```

The factor diagram shows that the five factors correspond nicely to the
5 types of questions:

``` r
fa.diagram(fit)
```

![](man/figures/README-fa-diagram-1.svg)<!-- -->

We can add the factor scores to the original data, in order to explore
group differences. Because we need complete data to calculate factor
scores, we first impute a single data set with mice:

``` r
data_imp <- mice::complete(mice::mice(bfi, 1, print = FALSE))

fct_scores <- data.frame(factor.scores(data_imp[, 1:25], fit)$scores)

data_imp <- data.frame(
  Gender        = factor(data_imp$gender),
  Extraversion  = fct_scores$MR1,
  Neuroticism   = fct_scores$MR2,
  Conscientious = fct_scores$MR3,
  Openness      = fct_scores$MR4,
  Agreeableness = fct_scores$MR5
)

levels(data_imp$Gender) <- c("Male", "Female")
```

Then we can visualize the group differences:

``` r
library(ggplot2)
library(tidyr)

data_imp2 <- tidyr::pivot_longer(data_imp, -Gender, "factor")

ggplot(data_imp2) +
  geom_density(aes(value, linetype = Gender)) +
  facet_wrap(~ factor, nrow = 2) +
  theme(legend.position = c(.9, .1))
```

![](man/figures/README-fa-group-comparison-1.svg)<!-- -->

## Further Reading

[1] Rubin D. B. Multiple imputation for nonresponse in surveys (2004).
John Wiley & Sons.

[2] Fieller, E. C. (1954). Some problems in interval estimation. Journal
of the Royal Statistical Society. Series B (Methodological): 175-185.

[3] Shao, J. & Sitter, R. R. (1996). Bootstrap for imputed survey data.
Journal of the American Statistical Association 91.435 (1996):
1278-1288. doi:
[10.1080/01621459.1996.10476997](https://dx.doi.org/10.1080/01621459.1996.10476997)


NEWS.md
# mifa 0.2.0

* Major code refactoring for readability and maintainability
* Vastly improved function signatures (breaking changes to v0.1.0)
* Added argument checks for all user-facing functions
* Add `cov_vars` argument for selecting variables to use for for imputations 
  but not for covariance matrix 
* Complete control over `mice()` via `...` arguments
* Better function returns, e.g. main function now returns an `S3` object of 
  class `mifa` which has nice `summary()` and `print()` generics
* Much improved documentation and examples 
* Added a `README.md` with installation and usage instructions
* Added a `{pkgdown}` site at <https://teebusch.github.io/mifa/>
* Added a `NEWS.md` file to track changes to the package.
* Remove unused dependencies
* Added continuous integration and unit tests


# mifa 0.1.0

* Initial release, accompanying the paper in Behavioral Research Methods 
  [(link)](https://doi.org/10.3758/s13428-017-1013-4)
* mifa 0.1.0 was originally released at <https://github.com/vahidnassiri/mifa/>
  but this repository appears to have been abandoned


tests/testthat.R
library(testthat)
library(mifa)

test_check("mifa")


tests/testthat/test-mifa.R
library(dplyr)

# The bfi data from the psych package contains answers to 5 personality items
# and 3 additional demographics for 2800 participants. The 25 columns with
# responses are already classified into 5 personality factors, as indicated by
# their names: A (Agreeableness), C (Conscientiousness), E (Extraversion),
# N (Neuroticism), O (Openness). The data contains some missing values
data_bfi <- psych::bfi[, 1:25]


test_that("mifa() returns have the expected type and shape", {
  m <- 2
  c <- ncol(data_bfi)
  res <- mifa(data_bfi, m = m, print = FALSE)

  expect_type(res, "list")
  expect_s3_class(res, "mifa")

  # cov_combined is a single matrix of type double
  expect_type(res$cov_combined, "double")
  expect_equal(dim(res$cov_combined), c(25, 25))

  # cov_imputations is a list of matrixes of type double
  expect_type(res$cov_imputations, "list")
  expect_true(all(sapply(res$cov_imputations, typeof) == "double"))
  expect_length(res$cov_imputations, m)
  expect_true(all(
    sapply(
      res$cov_imputations,
      function(x) all(dim(x) == c(c, c))
    )
  ))

  # var_explained is a data frame
  expect_s3_class(res$var_explained, "data.frame")
  expect_length(res$var_explained$n_pc, c)
  expect_true(all(res$var_explained$var_explained >= 0 &
    res$var_explained$var_explained <= 1))

  # CIs are not computed by default and return NULL
  expect_null(res$var_explained$ci_boot_lower)
  expect_null(res$var_explained$ci_boot_upper)
  expect_null(res$var_explained$ci_fieller_lower)
  expect_null(res$var_explained$ci_fieller_upper)
})


test_that("CIs returned by mifa() have the expected type and shape", {
  m <- 2
  c <- ncol(data_bfi)
  res <- mifa(data_bfi,
    n_pc = 3:5, ci = "both", n_boot = 2, m = m,
    maxit = 2, print = FALSE
  )
  cis <- res$var_explained


  expect_type(res, "list")
  expect_s3_class(res, "mifa")
  expect_s3_class(cis, "data.frame")

  expect_type(cis$ci_boot_lower, "double")
  expect_type(cis$ci_boot_upper, "double")
  expect_type(cis$ci_fieller_lower, "double")
  expect_type(cis$ci_fieller_upper, "double")

  expect_length(cis$ci_boot_lower, 3)
  expect_length(cis$ci_boot_upper, 3)
  expect_length(cis$ci_fieller_lower, 3)
  expect_length(cis$ci_fieller_upper, 3)
})


test_that("cov_vars argument removes variables", {
  m <- 2
  c <- ncol(data_bfi)
  c_sel <- ncol(dplyr::select(data_bfi, starts_with("O")))

  res <- mifa(data_bfi, cov_var = starts_with("O"), m = m, print = FALSE)

  expect_true(is.matrix(res$cov_combined))
  expect_type(res$cov_combined, "double")
  expect_length(res$cov_combined, c_sel^2)
  expect_equal(dim(res$cov_combined), c(c_sel, c_sel))
})


test_that("output is equivalent to that of original implementation", {
  # the essence of the original mifa algorithm, without cis
  mifa_old <- function(data, n.factor, M, maxit.mi = 5, method.mi = "pmm") {
    N <- dim(data)[1]
    imputed_mice <- mice::mice(data, m = M, maxit = maxit.mi,
                               method = method.mi, print = FALSE)
    method.levels.mi <- levels(imputed_mice$loggedEvents$meth)
    if ("constant" %in% method.levels.mi) {
      stop("Probably at least one column with constant observed part.")
    }
    comp.mice <- NULL
    mi.na <- rep(0, M)
    for (i in 1:M) {
      comp.mice[[i]] <- mice::complete(imputed_mice, i)
      mi.na[i] <- sum(is.na(comp.mice[[i]]))
    }
    while (sum(mi.na) > 0) {
      for (i in 1:M) {
        imp.tmp <- mice::mice(comp.mice[[i]], m = 1, maxit = maxit.mi,
                              method = method.mi, print = FALSE)
        comp.mice[[i]] <- mice::complete(imp.tmp, 1)
        mi.na[i] <- sum(is.na(comp.mice[[i]]))
      }
    }
    cov.mice.imp <- NULL
    prop.exp <- rep(0, M)
    for (i in 1:M) {
      cov.tmp <- cov(comp.mice[[i]])
      cov.mice.imp[[i]] <- cov.tmp
    }
    cov.mice     <- Reduce("+", cov.mice.imp) / M
    eig.cov.mice <- eigen(cov.mice)$values
    exp.var.mice <- (cumsum(eig.cov.mice) / sum(eig.cov.mice))[n.factor]

    list(
      cov.mice     = cov.mice,
      cov.mice.imp = cov.mice.imp,
      exp.var.mice = exp.var.mice
    )
  }

  m <- 2
  maxit <- 3

  set.seed(111)
  old <- mifa_old(data_bfi, M = m, maxit.mi = maxit)
  # latest version
  set.seed(111)
  new <- mifa(data_bfi, m = 2, maxit = maxit, print = FALSE)

  expect_equal(old$cov.mice, new$cov_combined)
  expect_equal(old$exp.var.mice, new$var_explained$var_explained,
               tolerance = 0.0001)
})


tests/testthat/test-mifa-ci.R
library(dplyr)

data_bfi <- psych::bfi[, 1:25]


test_that("mifa_ci_boot() returns right shape and type", {
  n_pc <- 3:5
  res <- mifa_ci_boot(data_bfi, n_pc = n_pc, n_boot = 3, maxit = 1, print = F)

  expect_s3_class(res, "data.frame")
  expect_equal(res$n_pc, n_pc)
  expect_type(res$lower, "double")
  expect_type(res$upper, "double")
  expect_equal(dim(res), c(3, 3))
})


test_that("using cov_vars argument produces different cis", {
  # can only test it indirectly, unfortunately
  m <- 2

  set.seed(123)
  res1 <- mifa_ci_boot(data_bfi, n_boot = 3, maxit = 1, print = F) # no selection

  set.seed(123) # use same seed for bootstrapping
  res2 <- mifa_ci_boot(data_bfi,
    cov_vars = starts_with("O"),
    n_boot = 3, maxit = 1, print = F
  ) # with selection

  expect_s3_class(res2, "data.frame")
  expect_false(identical(res1, res2))
  expect_equal(dim(res1), c(ncol(data_bfi), 3))
  expect_equal(dim(res2), c(ncol(dplyr::select(data_bfi, starts_with("O"))), 3))
})


test_that("mifa_ci_fieller() returns right shape and type", {
  n_pc <- 3:5
  m <- 2

  res_mifa <- mifa(data_bfi, n_pc = 2, m = m, maxit = 2, print = F)

  res <- mifa_ci_fieller(res_mifa$cov_imputations,
    n_pc = n_pc,
    N = nrow(data_bfi)
  )

  expect_s3_class(res, "data.frame")
  expect_equal(res$n_pc, n_pc)
  expect_type(res$lower, "double")
  expect_type(res$upper, "double")
  expect_equal(dim(res), c(3, 3))
})


test_that("combine_rubin() return has the right shape and type", {
  # make some input of the expected shape
  data <- na.omit(data_bfi)
  c <- ncol(data)
  param_imps <- list()
  cov_imps <- list()

  for (i in 1:5) {
    param_imps[[i]] <- eigen(cov(data))$values
    cov_imps[[i]] <- eigen(cov(data))$vectors
  }

  param_imps <- data.frame(Reduce(rbind, param_imps))

  # tests
  res <- combine_rubin(param_imps, cov_imps)

  expect_type(res, "list")
  expect_length(res, 3)

  expect_type(res$param_est, "double")
  expect_length(res$param_est, c)

  expect_type(res$cov_param, "double")
  expect_equal(dim(res$cov_param), c(c, c))

  expect_type(res$cov_between, "double")
  expect_equal(dim(res$cov_between), c(c, c))
})


R/mifa-ci.R
#' Bootstrap confidence intervals for explained variance
#'
#' Compute bootstrap confidence intervals for the proportion of explained
#' variance for the covariance of an incomplete data imputed using
#' multiple imputation.
#' For multiple imputation, Multivariate Imputation by Chained Equations
#' (MICE) from the [mice] package is used.
#'
#' This function uses the Shao and Sitter (1996) method to combine multiple
#' imputation and bootstrapping. The imputations are done using [mice::mice()].
#'
#' Normally, this function does not need to be called directly. Instead,
#' use `mifa(..., ci = "boot")`.
#'
#' @references
#' Shao, J. & Sitter, R. R. (1996). Bootstrap for imputed survey data.
#' Journal of the American Statistical Association 91.435 (1996): 1278-1288.
#' \doi{10.1080/01621459.1996.10476997}
#'
#' @param progress Logical. Whether to show progress bars for computation of
#' bootstrap confidence intervals. Default is FALSE.
#' @inheritParams mifa
#' @inheritDotParams mice::mice
#'
#' @seealso [mifa()], [mice::mice()]
#' @family mifa confidence intervals
#'
#' @return A data frame containing bootstrapped confidence intervals for
#' variance explained by different number of principal components.
#' @export
#' @examples
#' \donttest{
#' if(requireNamespace("psych")) {
#'   data <- psych::bfi[, 1:25]
#'   mifa_ci_boot(data, n_pc = 3:8, n_boot = 10, print = FALSE)
#' }
#' }
mifa_ci_boot <- function(data, cov_vars = dplyr::everything(), n_pc,
                         conf = .95, n_boot = 1000, progress = FALSE, ...) {
  # check and clean arguments
  checkmate::assert_data_frame(data, min.rows = 1, min.cols = 2)
  checkmate::assert_data_frame(dplyr::select(data, {{ cov_vars }}), min.cols = 2)
  checkmate::assert_number(conf, lower = 0, upper = 1)
  checkmate::assert_count(n_boot, positive = TRUE)
  checkmate::assert_flag(progress)

  n_cov_vars <- ncol(dplyr::select(data, {{ cov_vars }}))
  n_pc <- clean_n_pc(n_pc, n_cov_vars)

  # bootstrap samples and compute variance explained by principal components
  # (i.e. eigenvalues of eigenvectors)
  boot_eig <- matrix(0, n_cov_vars, n_boot)

  if (progress) {
    cat("\n\n  Computing bootstrap confidence intervals...\n")
    pb <- utils::txtProgressBar(min = 0, max = n_boot, initial = 0, style = 3)
  }
  for (i in 1:n_boot) {
    # draw bootstrap samples and impute until there are no constants
    repeat {
      boot_idx <- sample(nrow(data), replace = TRUE)
      data_boot <- data[boot_idx, ]
      try({
        imp <- stop_constants(mice_impute_once(data_boot, ...))
        break
      })
    }

    # extract imputed dataset, make sure no NAs are left
    data_imp <- mice::complete(imp)
    data_imp <- mice_impute_all_NA(data_imp, ...)

    # Select variables for calculation of covariance matrix
    data_imp <- dplyr::select(data_imp, {{ cov_vars }})

    # eigenvalues of covariance matrix
    boot_eig[, i] <- eigen(stats::cov(data_imp))$values

    if (progress) utils::setTxtProgressBar(pb, i)
  }

  # get confidence intervals for variance explained by  principal components
  # (i.e., eigenvalues of eigenvectors)
  var_expl <- t(apply(boot_eig, 2, cumsum)) / apply(boot_eig, 2, sum)
  probs_ci <- c((1 - conf) / 2, 1 - (1 - conf) / 2)
  boot_expl <- apply(var_expl, 2, stats::quantile, probs = probs_ci)
  boot_cis <- t(boot_expl)[n_pc, ]

  data.frame(
    n_pc = n_pc,
    lower = boot_cis[, 1],
    upper = boot_cis[, 2]
  )
}



#' Fieller confidence intervals for explained variance
#'
#' Computes parametric confidence intervals for proportion of explained
#' variance for given numbers of principal components using Fieller's method.
#' Note that by setting `ci = TRUE` in [mifa()], this confidence
#' interval can be computed as well.
#'
#' Normally, this function does not need to be called directly. Instead,
#' use `mifa(..., ci = "fieller")`.
#'
#' @references
#' Fieller, E. C. (1954). Some problems in interval estimation.
#' Journal of the Royal Statistical Society. Series B (Methodological): 175-185.
#'
#' @param cov_imps List containing the estimated covariance matrix within
#' each imputed data. One can use `cov_imputations` returned by [mifa()].
#' @param N A scalar specifying sample size.
#'
#' @inheritParams mifa
#'
#' @seealso [mifa()]
#' @family mifa confidence intervals
#'
#' @return A data frame containing confidence intervals for `n_pc` principal
#' components.
#' @export
#' @examples
#' \donttest{
#' if(requireNamespace("psych")) {
#'   data <- psych::bfi[, 1:25]
#'   mi <- mifa(data, print = FALSE)
#'   mifa_ci_fieller(mi$cov_imputations, n_pc = 3:8, N = nrow(data))
#' }
#' }
mifa_ci_fieller <- function(cov_imps, n_pc, conf = .95, N) {
  # check and clean arguments
  checkmate::assert_list(cov_imps, "matrix", any.missing = FALSE, min.len = 1)
  lapply(cov_imps, function(x) {
    checkmate::assert_matrix(x, ncols = nrow(x), any.missing = FALSE)
  })
  checkmate::assert_number(conf, lower = 0, upper = 1)
  checkmate::assert_count(N, positive = TRUE)

  n_cov_vars <- ncol(cov_imps[[1]])
  n_pc <- clean_n_pc(n_pc, n_cov_vars)

  # get eigenvalues
  m <- length(cov_imps)
  eig_imp <- matrix(0, m, dim(cov_imps[[1]])[1])

  for (i in 1:m) {
    eig_imp[i, ] <- eigen(cov_imps[[i]])$values
  }

  # compute Fieller confidence intervals for each number of principal components
  ci_fieller <- matrix(NA, length(n_pc), 2)
  for (i in seq_along(n_pc)) {
    try({
      ci_fieller[i, ] <- get_fieller_ci(eig_imp, n_pc[i], conf, N, m)
    })
  }

  out <- data.frame(cbind(n_pc, ci_fieller))
  colnames(out) <- c("n_pc", "lower", "upper")
  return(out)
}



#' Find the Fieller interval for each k
#'
#' This function is used by [mifa_ci_fieller()] to compute Fieller's confidence
#' intervals for each of the components of n_pc.
#'
#' @param eig_imp A matrix with each of its columns the eigenvalues of the
#' estimated covariance matrix for each imputed data.
#' @param m A scalar specifying number of multiple imputations.
#'
#' @inheritParams mifa_ci_fieller
#'
#' @return A vector of length 2, containing the lower and upper bounds of
#' estimated Fieller's interval.
#' @keywords internal
get_fieller_ci <- function(eig_imp, n_pc, conf, N, m) {

  # combine imputations
  eig_imp2 <- eig_imp^2
  mi_cov <- lapply(1:m, function(i) (2 / N) * diag(eig_imp2[i, ]))

  mi_comb <- combine_rubin(eig_imp, mi_cov)
  cov_lambda <- mi_comb$cov_param

  # compute Fieller's intervals
  P <- dim(cov_lambda)[1]
  A1 <- matrix(c(rep(1, n_pc), rep(0, P - n_pc)), 1, P)
  A <- matrix(1, 1, P)
  s11_comb <- (A1 %*% cov_lambda) %*% t(A1)
  s22_comb <- (A %*% cov_lambda) %*% t(A)
  s12_comb <- (A1 %*% cov_lambda) %*% t(A)
  S <- matrix(c(s11_comb, s12_comb, s12_comb, s22_comb), 2, 2)
  s11 <- S[1, 1]
  s22 <- S[2, 2]
  s12 <- S[1, 2]

  eigen.all <- sum(mi_comb$param_est)
  eigen.first <- sum(mi_comb$param_est[1:n_pc])

  C12 <- s11 / (eigen.first^2)
  C22 <- s22 / (eigen.all^2)
  r <- s12 / sqrt(s11 * s22)
  T.crit <- stats::qnorm(1 - ((1 - conf) / 2))
  A <- C12 + C22 - (2 * r * sqrt(C12 * C22))
  B <- (T.crit^2) * C12 * C22 * (1 - (r^2))

  if ((A - B) < 0) {
    stop("Computing Fieller CI is not possible. Try using more imputations.")
  }

  fieller1 <- eigen.first / eigen.all
  fieller2 <- 1 - (T.crit^2 * r * sqrt(C12 * C22))
  fieller3 <- T.crit * sqrt(A - B)
  fieller4 <- 1 - ((T.crit^2) * C22)

  c(
    lower = fieller1 * ((fieller2 - fieller3) / fieller4),
    upper = fieller1 * ((fieller2 + fieller3) / fieller4)
  )
}



#' Combine results from different imputations using Rubin's rules
#'
#' Applies Rubin's rules to combine estimates and
#' variance-covariance matrices from different imputations.
#'
#' @references
#' Rubin D. B. Multiple imputation for nonresponse in surveys (2004).
#' John Wiley & Sons.
#'
#' @param param_imps Matrix containing estimated parameters in each imputation
#' as its rows.
#' @param cov_imps List of estimated covariance matrices for each imputation.
#'
#' @return A list:
#' \describe{
#'   \item{param_est}{Vector of combined parameter estimates with the same length
#'   as columns in `param_imps`.}
#'   \item{cov_param}{Combined variance-covariance matrix of size n x n, where n
#'   is the number of columns in `param_imps`.}
#'   \item{between_cov}{Between imputations variance-covariance matrix of size
#'   n x n, where n is the number of columns in `param_imps`.}
#' }
#' @keywords internal
combine_rubin <- function(param_imps, cov_imps) {
  m <- length(cov_imps)

  est_diff <- scale(param_imps, center = TRUE, scale = FALSE)
  cov_param_sample1 <- lapply(1:m, function(i) est_diff[i, ] %*% t(est_diff[i, ]))
  cov_param_sample1 <- Reduce("+", cov_param_sample1) / (m - 1)

  cov_param_sample <- cov_param_sample1 * ((m + 1) / m)
  cov_param_mean <- Reduce(`+`, cov_imps) / m
  cov_param <- cov_param_mean + cov_param_sample

  list(
    param_est = apply(param_imps, 2, mean),
    cov_param = cov_param,
    cov_between = cov_param_sample1
  )
}


R/mice-helpers.R
#' Imputes a single dataset using mice
#'
#' @inheritParams mifa
#'
#' @return A data frame with missing values imputed.
#' @keywords internal
mice_impute_once <- function(data, ...) {
  args <- list(...)
  args$m <- 1
  args$data <- data
  imp <- do.call(mice::mice, args)
}


#' Sequential Imputation
#'
#' Sequential imputations in case that some of the columns are
#' not imputed due to collinearity, etc.
#'
#' This function will keep imputing the same data over and over
#' until there are no NA values left.
#'
#' @inheritParams mifa
#'
#' @return A data frame without NAs
#' @keywords internal
mice_impute_all_NA <- function(data, ...) {
  if (any(is.na(data))) {
    message(paste(
      "There are", sum(is.na(data)), "NAs left in the data after",
      "using MICE. Mifa will try to use sequential MICE to impute",
      "them..."
    ))
  }
  while (any(is.na(data))) {
    imp <- mice_impute_once(data, ...)
    data <- mice::complete(imp)
  }
  data
}


#' A wrapper for a call to mice that throws an error if there are
#' constants in the data
#'
#' @param f a call to [mice::mice()]
#' @return The result of calling the provided function or an Error if there is
#' a constant in one of the columns.
#' @keywords internal
stop_constants <- function(f) {
  imp <- eval(f)
  if ("constant" %in% levels(imp$loggedEvents$meth)) {
    stop("At least one column with constant observed part.")
  }
  imp
}


R/mifa.R
#' Get covariance matrix of incomplete data using multiple imputation
#'
#' Compute covariance matrix of incomplete data using multiple imputation.
#' For multiple imputation, *Multivariate Imputation by Chained Equations*
#' (MICE) from the [mice] package is used. The covariance matrices of the
#' imputed data sets are combined using Rubin's rules.
#'
#' The function also computes the variance explained by different numbers of
#' principal components and the corresponding Fieller (parametric) or bootstrap
#' (nonparametric) confidence intervals.
#'
#' @references
#' Nassiri, V., Lovik, A., Molenberghs, G., & Verbeke, G. (2018).
#' On using multiple imputation for exploratory factor analysis of incomplete
#' data. Behavioral Research Methods 50, 501–517.
#' \doi{10.3758/s13428-017-1013-4}
#'
#' @param data A data frame with missing values coded as `NA`.
#' @param cov_vars Variables in `data` for which to calculate the covariance
#' matrix. Supports (tidy selection)[dplyr::select()]. This allows to
#' select variables that are used for the imputations of missing values, but not
#' the calculations of the covariance matrix. This is especially useful when
#' there are categorical predictors that can improve the imputation of the
#' response variables, but for which covariance cannot be calculated.
#' By default, all variables in `data` are used for both, the imputation and
#' the covariance matrix. Note: Variables and rows used for the imputation, as
#' well as the method for imputation can be configured using the `...`.
#' See also [mice::mice()].
#' @param n_pc Integer or integer vector indicating number of principal
#' components (eigenvectors) for which explained variance (eigenvalues) should
#' be obtained and for which confidence intervals should be computed.
#' Defaults to all principal components, i.e., the number of variables in the
#' data.
#' @param conf Confidence level for constructing confidence intervals. The
#' default is `.95` that is, 95% confidence intervals.
#' @param n_boot Number of bootstrap samples to use for bootstrapped confidence
#' intervals. The default is 1000.
#' @param ci A character string indicating which types of confidence intervals
#' should be constructed for the variance explained by the principal
#' components. If `"boot"`, `"fieller"`, or `"both"`, the corresponding
#' intervals are computed. If `FALSE` (the default) no confidence intervals will
#' be computed. The components for which confidence intervals should be computed
#' can be set with `n_pc`. See [mifa_ci_boot()] and [mifa_ci_fieller()] for
#' details about the two methods.
#' @inheritDotParams mice::mice
#'
#' @seealso [mifa_ci_boot()], [mifa_ci_fieller()], [mice::mice()]
#'
#' @return A list:
#' \describe{
#'   \item{cov_combined}{The estimated covariance matrix of the
#'   incomplete data, based on the combined covariance matrices of imputed
#'   data sets.}
#'   \item{cov_imputations}{A list containing the estimated covariance matrixes
#'   for all imputed data sets.}
#'   \item{var_explained}{A data frame containing the estimated proportions of
#'   explained variance for each of specified `n_pc` components. Depending o
#'   n `ci`, it will also contain the estimated Fieller's (parametric) and/or
#'   bootstrap (nonparametric) confidence interval for the proportion of
#'   variance explained by the different numbers of principal components defined
#'   by `n_pc`.}
#'   \item{mids}{Object of type [mice::mids]. This is the results of
#'   the multiple imputation step for the covariance matrix. Can be useful for
#'   diagnosing the multiple imputations.}
#' }
#' @export
#' @examples
#' \donttest{
#' if(requireNamespace("psych")) {
#'   data <- psych::bfi
#'   mifa(data, cov_vars = -c(age, education, gender), ci = "fieller", print = FALSE)
#' }
#' }
mifa <- function(data, cov_vars = dplyr::everything(), n_pc, ci = FALSE,
                 conf = .95, n_boot = 1000, ...) {
  # check and clean arguments
  checkmate::assert_data_frame(data, min.rows = 1, min.cols = 2)
  checkmate::assert_data_frame(dplyr::select(data, {{ cov_vars }}), min.cols = 2)
  if (ci != FALSE) {
    checkmate::assert_choice(ci, c(FALSE, "both", "boot", "fieller"))
  }
  checkmate::assert_number(conf, lower = 0, upper = 1)
  checkmate::assert_count(n_boot, positive = TRUE)

  n_cov_vars <- ncol(dplyr::select(data, {{ cov_vars }}))
  n_pc <- clean_n_pc(n_pc, n_cov_vars)

  # create imputed datasets, make sure no NAs are left
  imp <- stop_constants(mice::mice(data, ...))
  data_imps <- mice::complete(imp, "all")
  data_imps <- lapply(data_imps, function(x) mice_impute_all_NA(x, ...))

  # Select variables for calculating covariance matrix
  data_imps <- lapply(data_imps, function(d) dplyr::select(d, {{ cov_vars }}))

  # estimate covariance matrix of imputed values and combine estimates
  cov_imps <- lapply(data_imps, stats::cov)
  cov_comb <- Reduce("+", cov_imps) / length(cov_imps)

  # get proportion of explained variance for different number of pr. components
  cov_comb_eigen <- eigen(cov_comb)$values

  var_expl <- data.frame(
    n_pc = n_pc,
    var_explained = (cumsum(cov_comb_eigen) / sum(cov_comb_eigen))[n_pc]
  )

  # add confidence intervals for variance explained
  if ("boot" %in% ci || "both" %in% ci) {
    ci_boot <- mifa_ci_boot(data, {{ cov_vars }}, n_pc, conf, n_boot, ...)
    var_expl <- dplyr::bind_cols(
      var_expl,
      ci_boot_lower = ci_boot$lower,
      ci_boot_upper = ci_boot$upper
    )
  }
  if ("fieller" %in% ci || "both" %in% ci) {
    ci_fieller <- mifa_ci_fieller(cov_imps, n_pc, conf, nrow(data))
    var_expl <- dplyr::bind_cols(
      var_expl,
      ci_fieller_lower = ci_fieller$lower,
      ci_fieller_upper = ci_fieller$upper
    )
  }

  structure(
    list(
      cov_combined = cov_comb,
      cov_imputations = cov_imps,
      var_explained = var_expl,
      mids = imp
    ),
    class = "mifa"
  )
}



#' @export
summary.mifa <- function(object, ...) {
  mi <- object
  cat(sprintf(
    "Imputed covariance matrix of %i variables\n\n",
    ncol(mi$cov_combined)
  ))

  last_to_print <- function(x, max_width = getOption("width")) {
    widths <- pmax(nchar(names(x)), nchar(x))
    Position(function(x) x > max_width, cumsum(widths), nomatch = length(x))
  }

  # Details about data

  vars_cov <- colnames(mi$cov_combined)
  n_missing <- colSums(is.na(mi$mids$data[, vars_cov]))
  widths <- pmax(nchar(names(n_missing)), nchar(n_missing))
  n_print <- last_to_print(n_missing)

  cat("Variable:  ", sprintf("%*s", widths[1:n_print], names(n_missing)[1:n_print]), "\n")
  cat("N Imputed: ", sprintf("%*i", widths[1:n_print], n_missing[1:n_print]))

  if (n_print < length(vars_cov)) {
    cat(sprintf(" ...\n...and %i more variables", length(vars_cov) - n_print))
  }

  # Details on imputation

  cat(sprintf("\n\nNumber of MICE imputations: %i", length(mi$cov_imputations)))

  vars_imp <- colnames(mi$mids$data)
  vars_diff <- setdiff(vars_imp, vars_cov)

  if (length(vars_diff) > 0) {
    cat(sprintf("\nAdditional variables used for imputations:\n"))

    n_print <- last_to_print(vars_diff)

    cat(sprintf("%s", vars_diff[1:n_print]))
    if (n_print < length(vars_diff)) {
      cat(sprintf("... \n...and %i more variables", length(vars_diff) - n_print))
    }
  }

  # variance expplained

  if (is.null(mi$var_explained$ci_fieller_lower)) {
    ci_fieller <- NULL
  } else {
    ci_fieller <- sprintf(
      "[%.2f, %.2f]",
      mi$var_explained$ci_fieller_lower,
      mi$var_explained$ci_fieller_upper
    )
  }

  if (is.null(mi$var_explained$ci_boot_lower)) {
    ci_boot <- NULL
  } else {
    ci_boot <- sprintf(
      "[%.2f, %.2f]",
      mi$var_explained$ci_boot_lower,
      mi$var_explained$ci_boot_upper
    )
  }

  k_width <- max(nchar(mi$var_explained$n_pc))
  cat("\n\nCumulative proportion of variance explained by n principal components:")
  cat(sprintf("\n\n%*s  prop  ", k_width, "n"))
  if (!is.null(ci_fieller)) cat("Fieller CI    ")
  if (!is.null(ci_boot)) cat("Bootstrap CI")

  cat(
    "\n",
    sprintf(
      "%*i %5.2f  %s  %s\n",
      k_width,
      mi$var_explained$n_pc,
      mi$var_explained$var_explained,
      if (is.null(ci_fieller)) "" else ci_fieller,
      if (is.null(ci_boot)) "" else ci_boot
    ),
    "\n",
    sep = ""
  )
}



#' @export
print.mifa <- function(x, ...) {
  summary(x)
}
